### ISSUE TITLE
Pluralisation of @NLparameters

### ISSUE THREAD

**Issue by path-doc** [jump-dev/JuMP.jl#2593]
Hi following on from the discourse thread:
https://discourse.julialang.org/t/pluralisation-of-nlparameter/61060

I'd like to request pluralisation of `@NLparameter`.

Thanks!

**Comment by odow** [comment]
It may be as simple as adding to this list:
https://github.com/jump-dev/JuMP.jl/blob/80626541cc19005d1cb48958531d2ec71e8b7bf7/src/macros.jl#L948-L956

Then it just requires docs and tests.

**Comment by Pramodh-G** [comment]
Hi! I'm a beginner to the JuMP ecosystem (Julia in general too), would it be good if I could work on this with your help?

**Comment by odow** [comment]
Sure. The steps are
- Develop the package `] dev JuMP`
- Go to `~/.Julia/dev/JuMP`
- Make a change to `src/macros.jl` by adding `(:NLparameters, Symbol("@NLparameter")),` to the list
- Open Julia and test that it works.
- Make a PR
- We also need to add some tests and documentation
  -  For the docs, copy https://github.com/jump-dev/JuMP.jl/blob/b3687f39f34420d8246318cb42a95a782b327f33/src/macros.jl#L1114-L1131 and add it to the API Reference https://github.com/jump-dev/JuMP.jl/blob/master/docs/src/reference/nlp.md
  - For the tests, add a new test set here: https://github.com/jump-dev/JuMP.jl/blob/b3687f39f34420d8246318cb42a95a782b327f33/test/nlp.jl#L708-L712

Let me know if you have questions on specific steps.

**Comment by path-doc** [comment]
I tried this, but I get the following error (using same format as for @variables, but the error is the same if I remover the base_name part):

julia> using JuMP
[ Info: Precompiling JuMP [4076af6c-e467-56ae-b986-b466b2749572]
julia> using Ipopt
julia> m = Model(Ipopt.Optimizer)
A JuMP Model
Feasibility problem with:
Variables: 0
Model mode: AUTOMATIC
CachingOptimizer state: EMPTY_OPTIMIZER
Solver name: Ipopt

julia> @NLparameter(m, a1 == 2)
"Reference to nonlinear parameter #1"

julia> @NLparameters(m,
                   begin
                   a == 1,
                   b == 2
                   end)
ERROR: LoadError: At REPL[29]:3: `@NLparameter(m, a == 1, b == 2)`: Too many arguments.
Stacktrace:
 [1] error(::String, ::String)
   @ Base ./error.jl:42
 [2] _macro_error(macroname::Symbol, args::Tuple{Symbol, Expr, Expr}, source::LineNumberNode, str::String)
   @ JuMP ~/.julia/dev/JuMP/src/macros.jl:1367
 [3] (::JuMP.var"#_error#122"{LineNumberNode, Symbol, Expr})(str::String)
   @ JuMP ~/.julia/dev/JuMP/src/macros.jl:2116
 [4] var"@NLparameter"(__source__::LineNumberNode, __module__::Module, m::Any, ex::Any, extra::Vararg{Any, N} where N)
   @ JuMP ~/.julia/dev/JuMP/src/macros.jl:2120
in expression starting at REPL[29]:3

**Comment by odow** [comment]
Two things:
1. You can format your code by enclosing it in triple-backticks
````
```Julia
Code goes here
```
````
2. Remove the comma after `a == 1,` :)

You should add a test like 
```Julia
@testset "NLparameters" begin
    model = Model()
    @NLparameters(model, begin
        a == 1
        b[i=1:2] == i
    end)
    @test value(a) == 1
    @test value.(b) == [1, 2]
end
```

**Comment by path-doc** [comment]
Thanks @odow. Yes, that works. 

It looks like @Pramodh-G has fixed this. 

I appreciate this is a separate matter, but I ended up with that comma there after trying to follow the format of @variables and give the parameters a base_name. This is not currently possible, yet the ability to name parameters would be useful. Should I raise this as an issue?

**Comment by odow** [comment]
Closed by #2619
